### PR TITLE
avoid init useless DeltaMergeStore

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/5665

Problem Summary: Some useless stores (most in mysql database) will be initialized when update `AsynchronousMetrics`. And it consumes unnecessary resource and make it harder to do end-to-end tests for data leak problem because these tables cannot be dropped.

### What is changed and how it works?
Don't initialize any `DeltaMergeStore` when update `AsynchronousMetrics`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
1. deploy a new cluster and check whether there is any data generated for the tables in mysql database;
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Interpreters/AsynchronousMetrics.cpp [hunk 1]
```diff
@@ -177,11 +177,14 @@ void AsynchronousMetrics::update()
 
                 if (auto dt_storage = std::dynamic_pointer_cast<StorageDeltaMerge>(table); dt_storage)
                 {
-                    auto stat = dt_storage->getStore()->getStat();
-                    calculateMax(max_dt_stable_oldest_snapshot_lifetime, stat.storage_stable_oldest_snapshot_lifetime);
-                    calculateMax(max_dt_delta_oldest_snapshot_lifetime, stat.storage_delta_oldest_snapshot_lifetime);
-                    calculateMax(max_dt_meta_oldest_snapshot_lifetime, stat.storage_meta_oldest_snapshot_lifetime);
-                    calculateMax(max_dt_background_tasks_length, stat.background_tasks_length);
+                    if (auto store = dt_storage->getStoreIfInited(); store)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Interpreters/AsynchronousMetrics.cpp:180]:
```suggestion
                    if (auto store = dt_storage->getStoreIfInited())
```

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 2]
```diff
@@ -982,21 +982,25 @@ std::vector<BlobFileId> BlobStore::getGCStats()
...
-                                                stat->id,
-                                                stat->sm_valid_size,
-                                                stat->sm_valid_rate));
+                    if (unlikely(stat->sm_valid_rate != 0))
+                    {
+                        throw Exception(fmt::format("Current blob is empty, but valid rate is not 0. [blob_id={}][valid_size={}][valid_rate={}]",
+                                                    stat->id,
+                                                    stat->sm_valid_size,
+                                                    stat->sm_valid_rate));
+                    }
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:994]:
```suggestion
                    RUNTIME_CHECK(stat->sm_valid_rate == 0, Exception(fmt::format("Current blob is empty, but valid rate is not 0. [blob_id={}][valid_size={}][valid_rate={}]",
                                                    stat->id,
                                                    stat->sm_valid_size,
                                                    stat->sm_valid_rate));
```

**Review by Lloyd-Pottiger — APPROVED** [review]
others LGTM

File: dbms/src/Interpreters/AsynchronousMetrics.cpp (hunk 1 above)
Comment (lidezhu) [reply dbms/src/Interpreters/AsynchronousMetrics.cpp:180]: It seems we always do similar check in this way. I prefer to keep the original way to do the check.

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 3]
```diff
@@ -982,21 +982,22 @@ std::vector<BlobFileId> BlobStore::getGCStats()
...
+                if (stat->sm_total_size != 0)
                 {
-                    throw Exception(fmt::format("Current blob is empty, but valid rate is not 0. [blob_id={}][valid_size={}][valid_rate={}]",
-                                                stat->id,
-                                                stat->sm_valid_size,
-                                                stat->sm_valid_rate));
+                    RUNTIME_CHECK(stat->sm_valid_rate == 0, Exception(fmt::format("Current blob is empty, but valid rate is not 0. [blob_id={}][valid_size={}][valid_rate={}]",
+                                                    stat->id,
+                                                    stat->sm_valid_size,
+                                                    stat->sm_valid_rate));
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:991]:
```suggestion
                    RUNTIME_CHECK_MSG(stat->sm_valid_rate == 0, "Current blob is empty, but valid rate is not 0. [blob_id={}][valid_size={}][valid_rate={}]",
                                                    stat->id,
                                                    stat->sm_valid_size,
                                                    stat->sm_valid_rate);
```

File: dbms/src/Interpreters/AsynchronousMetrics.cpp (hunk 1 above)
Comment (JaySon-Huang) [reply dbms/src/Interpreters/AsynchronousMetrics.cpp:180]: Using the return result of an assign statement is not recommended, it introduce bug like `if (a == b) / if (a = b)`.
Using `if (auto a = b; a)` or `if (auto a = b; a != nullptr)` is more explicit

Comment (Lloyd-Pottiger) [reply dbms/src/Interpreters/AsynchronousMetrics.cpp:180]: Get it, thank you

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 4]
```diff
@@ -982,21 +982,22 @@ std::vector<BlobFileId> BlobStore::getGCStats()
             // Avoid divide by zero
             if (right_margin == 0)
             {
-                if (unlikely(stat->sm_valid_rate != 0))
+                // Ignore the BlobFile if it is already truncated
+                if (stat->sm_total_size != 0)
```
[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/BlobStore.cpp:986]: Yes, I considered to write this way previously.
But 
```
if (stat->sm_total_size != 0) {
  RUNTIME_CHECK(...)
  ...
}
continue
```
seems a little clearer than
```
if (stat->sm_total_size == 0) {
  continue;
}
RUNTIME_CHECK(...)
continue;
```
?

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 5]
```diff
@@ -982,21 +982,19 @@ std::vector<BlobFileId> BlobStore::getGCStats()
...
                 {
-                    throw Exception(fmt::format("Current blob is empty, but valid rate is not 0. [blob_id={}][valid_size={}][valid_rate={}]",
-                                                stat->id,
-                                                stat->sm_valid_size,
-                                                stat->sm_valid_rate));
+                    RUNTIME_CHECK_MSG(stat->sm_valid_rate == 0, "Current blob is empty, but valid rate is not 0. [blob_id={}][valid_size={}][valid_rate={}]", stat->id, stat->sm_valid_size, stat->sm_valid_rate);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:988]: `sm_valid_rate` is a double, comparing it directly to 0 may fail. Should we compare stat->sm_valid_rate is closed to 0.0 by a epsilon?

File: dbms/src/Storages/Page/V3/BlobStore.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply dbms/src/Storages/Page/V3/BlobStore.cpp:986]: Well, this is not a blocking comment

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 6]
```diff
@@ -985,10 +985,7 @@ std::vector<BlobFileId> BlobStore::getGCStats()
                 // Ignore the BlobFile if it is already truncated
                 if (stat->sm_total_size != 0)
                 {
-                    RUNTIME_CHECK_MSG(stat->sm_valid_rate == 0, "Current blob is empty, but valid rate is not 0. [blob_id={}][valid_size={}][valid_rate={}]",
-                                                    stat->id,
-                                                    stat->sm_valid_size,
-                                                    stat->sm_valid_rate);
+                    RUNTIME_CHECK_MSG(stat->sm_valid_rate == 0, "Current blob is empty, but valid rate is not 0. [blob_id={}][valid_size={}][valid_rate={}]", stat->id, stat->sm_valid_size, stat->sm_valid_rate);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:988]:
```suggestion
                    RUNTIME_CHECK_MSG(stat->sm_valid_rate == 0, "Current blob is empty, but valid rate is not 0. [blob_id={}] [valid_size={}] [valid_rate={}]", stat->id, stat->sm_valid_size, stat->sm_valid_rate);
```

File: dbms/src/Storages/Page/V3/BlobStore.cpp (hunk 5 above)
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/BlobStore.cpp]: Use `sm_valid_size` instead.

File: dbms/src/Storages/Page/V3/BlobStore.cpp (hunk 6 above)
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/BlobStore.cpp]: Fixed.

**Review by JaySon-Huang — APPROVED** [review]
lgtm

**Pull request merged**